Dynavity/Dynavity/view/graph/GraphView.swift

```diff
@@ -0,0 +1,118 @@
+    private static let zoomScaleRange: ClosedRange<CGFloat> = 0.2...2.5
+
+    @StateObject var viewModel = GraphMapViewModel()
+
+    // For viewport dragging gesture
+    @State var originOffset: CGPoint = .zero
+    @State var dragOffset: CGSize = .zero
+
+    // For viewport zooming
+    @State var zoomScale: CGFloat = 1.0
+    @State var previousZoomScale: CGFloat = 1.0
+
+    @Binding var searchQuery: String
+
+    var body: some View {
+        GeometryReader { geometry in
+            ZStack {
+                Rectangle().fill(Color.UI.base5)
+                graphView
+            }
+            .drawingGroup(opaque: true, colorMode: .extendedLinear)
+            .gesture(getViewportDragGesture(viewportSize: geometry.size))
+            .gesture(viewportMagnificationGesture)
+        }
+    }
+
+    var graphView: some View {
+        Group {
+            nodesView
+                .zIndex(.infinity)
+            edgesView
+        }
+        .scaleEffect(self.zoomScale)
+        .offset(x: self.originOffset.x + self.dragOffset.width,
+                y: self.originOffset.y + self.dragOffset.height)
+        .animation(.easeIn)
+    }
+
+    var nodesView: some View {
+        ZStack {
+            ForEach(viewModel.getNodes(), id: \.id) { node in
+                NodeView(label: node.name, isHighlighted: doesInputMatchSearchQuery(input: node.name))
+                    .offset(x: node.position.x, y: node.position.y)
+            }
+        }
+    }
+
+    var edgesView: some View {
+        ZStack {
+            ForEach(viewModel.getEdges(), id: \.self) { edge in
+                EdgeView(start: edge.source.position, end: edge.destination.position)
+                    .stroke()
+            }
+        }
+    }
+}
+
+// MARK: Gestures
+extension GraphView {
+    private func getViewportDragGesture(viewportSize: CGSize) -> some Gesture {
+        DragGesture()
+            .onChanged { value in
+                // Dragging canvas instead of a node
+                if viewModel.selectedNode == nil {
+                    viewModel.hitTest(tapPos: value.startLocation,
+                                      viewportSize: viewportSize,
+                                      viewportZoomScale: zoomScale,
+                                      viewportOriginOffset: originOffset)
+                    self.dragOffset = value.translation
+                } else {
+                    viewModel.handleNodeDragChange(value, viewportZoomScale: zoomScale)
+                }
+            }
+            .onEnded { value in
+                // Dragged a canvas instead of a node
+                if viewModel.selectedNode == nil {
+                    self.dragOffset = .zero
+                    self.originOffset += value.translation
+                } else {
+                    viewModel.handleNodeDragEnd(value, viewportZoomScale: zoomScale)
+                }
+            }
+    }
+
+    private var viewportMagnificationGesture: some Gesture {
```

Comment:
Oops, I knew something didn't feel right. Just didn't know what exactly.
Unfortunately, I think this fix isn't something easy since MagnificationGesture doesn't give the location. I'll fix this some other time (hopefully) if that's OK